File: chapter1/1.py
'''
Implement an algorithm to determine if a
string has all unique characters. What if 
you cannot use additional data structures?

'''

def is_string_unique(str):
	character_set = set()
	for c in str:
		if c in character_set:
			return False
		else:
			character_set.add(c)
	return True


def main():
	print(is_string_unique("herlo"))

if __name__ == "__main__":
	main()

File: chapter1/c++/1.cpp
/*
 * Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data
 * structures?
 */
#include <bits/stdc++.h>
using namespace std;

bool isStringUnique(string s) {
	set<char> charset;
	for(int i=0; i<s.size(); i++) {
		if(charset.find(s[i])!=charset.end())
			return false;
		else
			charset.insert(s[i]);
	}
	return true;
}

bool isStringUniqueNoAdditionalDataStructures(string s) {
	for(int i=0; i<s.size(); i++) {
		for(int j=0; j<s.size(); j++) {
			if(i!=j)
				if(s[i]==s[j]) return false;
		}
	}
	return true;
}

int main() {
	cout << isStringUnique("hello") << endl;
	cout << isStringUniqueNoAdditionalDataStructures("hello") << endl;
}
